src/test/java/com/commercetools/sync/sdk2/products/helpers/ProductBatchValidatorTest.java

```diff
@@ -0,0 +1,540 @@
+import com.commercetools.api.models.type.TypeResourceIdentifierBuilder;
+import com.commercetools.sync.sdk2.commons.helpers.DefaultCurrencyUnits;
+import com.commercetools.sync.sdk2.customobjects.helpers.CustomObjectCompositeIdentifier;
+import com.commercetools.sync.sdk2.products.ProductSyncOptions;
+import com.commercetools.sync.sdk2.products.ProductSyncOptionsBuilder;
+import com.fasterxml.jackson.databind.node.JsonNodeFactory;
+import com.fasterxml.jackson.databind.node.NullNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.UUID;
+import javax.annotation.Nonnull;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class ProductBatchValidatorTest {
+  private List<String> errorCallBackMessages;
+  private ProductSyncOptions syncOptions;
+  private ProductSyncStatistics syncStatistics;
+
+  @BeforeEach
+  void setup() {
+    errorCallBackMessages = new ArrayList<>();
+    final ProjectApiRoot ctpClient = mock(ProjectApiRoot.class);
+    syncOptions =
+        ProductSyncOptionsBuilder.of(ctpClient)
+            .errorCallback(
+                (exception, oldResource, newResource, updateActions) -> {
+                  errorCallBackMessages.add(exception.getMessage());
+                })
+            .build();
+    syncStatistics = mock(ProductSyncStatistics.class);
+  }
+
+  @Test
+  void getReferencedProductKeys_WithNullAttributes_ShouldReturnEmptySet() {
+    assertThat(getReferencedProductKeys(ProductVariantDraftBuilder.of().build())).isEmpty();
+  }
+
+  @Test
+  void getReferencedProductKeys_WithNoAttributes_ShouldReturnEmptySet() {
+    assertThat(getReferencedProductKeys(ProductVariantDraftBuilder.of().attributes().build()))
+        .isEmpty();
+  }
+
+  @Test
+  void getReferencedProductKeys_WithANullAttribute_ShouldReturnEmptySet() {
+    assertThat(
+            getReferencedProductKeys(
+                ProductVariantDraftBuilder.of().attributes(singletonList(null)).build()))
+        .isEmpty();
+  }
+
+  @Test
+  void getReferencedProductKeys_WithAProductRefAttribute_ShouldReturnEmptySet() {
+    final Attribute productReferenceSetAttribute =
+        getReferenceSetAttributeDraft(
+            "foo", getProductReferenceWithId("foo"), getProductReferenceWithId("bar"));
+    final Attribute productReferenceAttribute =
+        AttributeBuilder.of().name("foo").value(getProductReferenceWithId("foo")).build();
+
+    final List<Attribute> attributes =
+        asList(null, productReferenceAttribute, productReferenceSetAttribute);
+
+    final ProductVariantDraft variantDraft =
+        ProductVariantDraftBuilder.of().attributes(attributes).build();
+    assertThat(getReferencedProductKeys(variantDraft)).containsExactlyInAnyOrder("foo", "bar");
```

Comment:
I wonder if this test is correct. 
You mock a ProductReference with ID calling `getProductReferenceWithId("bar")` and then you assert that `getReferencedProductKeys(variantDraft)` contains the IDs but not the keys...

Reply:
The test is correct. Indeed the naming is a bit weird. The old test is the same: https://github.com/commercetools/commercetools-sync-java/blob/89a4a0f194fb0e1666ffc9720c7cbf845f84c640/src/test/java/com/commercetools/sync/products/helpers/ProductBatchValidatorTest.java#L91

It seems like the problem is in the method [`getReferencedProductKeys`](https://github.com/commercetools/commercetools-sync-java/blob/89a4a0f194fb0e1666ffc9720c7cbf845f84c640/src/main/java/com/commercetools/sync/products/helpers/ProductBatchValidator.java#L248) - this method gets the [referenced product IDs](https://github.com/commercetools/commercetools-sync-java/blob/89a4a0f194fb0e1666ffc9720c7cbf845f84c640/src/main/java/com/commercetools/sync/products/helpers/ProductBatchValidator.java#L282), not the keys. So it seems like a wrong naming here.

As I was not involved in the development of this tool, I don't why it is named like this. However, this is a public method, so we should not change the name as this could be a breaking change for a customer.